### ISSUE TITLE
Conflict in NaiveSyncBatchNorm

### ISSUE THREAD

**Issue by fanq15** [facebookresearch/detectron2#1807]
In [NaiveSyncBatchNorm](https://github.com/facebookresearch/detectron2/blob/123366fe7fd412963bee6404c69c1e06a5cdf06c/detectron2/layers/batch_norm.py#L219), `B=0` so there should not be gradient w.r.t input.

**Comment by ppwwyyxx** [comment]
When input is empty the gradient w.r.t. input should be an empty tensor.
This is consistent with other layers in pytorch.

**Comment by fanq15** [comment]
When I use NaiveSyncBatchNorm with stats_mode=N and the input is empty, the training gets stuck

 and the GPU utils are 100%.

**Comment by ppwwyyxx** [comment]
If you need help to solve an unexpected issue you observed, please include details following the ["Unexpected behaviors" issue template](https://github.com/facebookresearch/detectron2/issues/new/choose).

The following script does not get stuck in my environment if I set one or two `N` to zero.
```python
import torch
from torch import nn
from torch.nn.parallel import DistributedDataParallel

from detectron2.engine import launch
from detectron2.utils import comm
from detectron2.layers.batch_norm import NaiveSyncBatchNorm

def main():
    C = 3
    bn = NaiveSyncBatchNorm(C, stats_mode="N")
    model = nn.Sequential(nn.Conv2d(C, C, 3, padding=1), bn)
    model = model.cuda()

    model = DistributedDataParallel(
        model, device_ids=[comm.get_local_rank()], broadcast_buffers=False
    )
    optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

    for _ in range(3):
        if comm.get_rank() == 0:
            N, H, W = 0, 3, 4
            data = torch.randn(N, C, H, W)
        else:
            N, H, W = 1, 5, 6
            data = torch.randn(N, C, H, W)
            # print(data.mean(dim=[0, 2, 3]))

        out = model(data.cuda())
        if N == 0:
            assert out.numel() == 0
        print(comm.get_rank(), bn.running_mean)
        print(comm.get_rank(), bn.running_var)
        optimizer.zero_grad()
        out.sum().backward()
        optimizer.step()

if __name__ == "__main__":
    launch(main, 2, dist_url="auto")
```

**Comment by fanq15** [comment]
My problem occurs in the multiple GPU environment. I use 4 GPUs to train my model and sometimes there will be empty input in one or more GPUs. Then the training gets stuck. And it performs well when I use the standard torch.nn.BatchNorm, but when use the nn.SyncBatchNorm or NaiveSyncBatchNorm, the training will get stuck. I thought the NaiveSyncBatchNorm(C, stats_mode="N") can solve this problem, but it didn't.
